### ISSUE TITLE
Focus / resize / positioning problem (summery)

### ISSUE THREAD

**Issue by fnagel** [fnagel/jquery-ui#147]
This is a summery of a couple of issues describing more or less the same problem: on (first) select (after resizing the window) the browser is jumping. 

_Issues involved (please see here for more precise description of the problem):_
#94
#96
#128
#137
#142
#154
#160

_Here is what we know so far:_
Problem seems to be related to the focus() in method open, the resize event in _create method and the calculation of the offset in relation to the overflow. The focus() is needed to receive the correct offset within refreshPosition as scrollTop and offset().top and position().top calculate wrong values without focus(). The mentioned resize event fires refreshPosition and gets pretty high and wrong values. Im not sure why as the timeout event gets correct values. 

_Solution_
Ahem. I disabled the resize event and it seems to work for now. Ive tested in FF5, IE8, IE7, IE6 and latest Chrome. 
I have no clue why this resize event has been implemented and we struggled with proxy handling. Please help me to test if the resize event was necessary.
I also noticed that at least (eventually IE8 on slow machines? see #94) IE7 needs a additional refreshPosition.

Please note: lately some other changes were made which could have improved the behavior of the widget: #80 and https://github.com/fnagel/jquery-ui/commit/9c9425ac48378d7357caebdf52fb2ec98e069b8b
This involves CSS!

_Problems_
No new ones. But IE7 and IE8 still have some flickering when opening first time and using maxHeight (overflow).

Please test the latest commit and give feedback!

**Comment by 422** [comment]
Thanks Felix, as always you are a star :)

**Comment by fnagel** [comment]
Thanks :-) It was a pain in the ass to sum up all the info and feedback, do testing and finally notice that the resize event is part of the problem.

Please note that I just pushed my local branch: https://github.com/fnagel/jquery-ui/commit/eaef46805aec411f80ff58c720e3766309997815

**Comment by fnagel** [comment]
Push! 

Please give some feedback!

**Comment by fnagel** [comment]
Please give some feedback!

**Comment by Ninkasi** [comment]
Disabling the resize event is not the answer here. It's causing the opened menu to freeze on the position, where it was before the resize event occured. 
I believe that the real problem lies within href="#", which is equivalent to "_top" thus causes a scroll to the top of the page.
Another thing i noticed was in "open"-method, where selected menu option is focused before refreshing positioning.
This seems cause jumping if the list of options is long (in my case i had selectmenu with 30+ options).

**Comment by fnagel** [comment]
Hey @Ninkasi and thanks for your contribution. Its really appreciated!

I commented your comment at the pull request.

**Comment by fnagel** [comment]
Possible fix here: #154

I will take a look asap.

**Comment by fnagel** [comment]
Commented on #154

Still no progress :-(

**Comment by fnagel** [comment]
Ive added a close event when resizing the window. Should be a nice workaround. 

Closing this issue as should been resolved. Please reopen new issues if problem persists.

**Comment by thg2k** [comment]
greetings,
today with current git version and chrome 14.0.835.202 i experienced the same problem described in this ticket. With the JS debugger i tracked down the problem to the .focus() call in open method, and the pointer in the comment sent me here.
I think this issue is not resolved. Simply calling _refreshPosition() _before_ the focus() call for me fixes the problem both on FF 7.0.1 and chrome 14.0.835.202
So what did i break?

**Comment by fnagel** [comment]
Im must confess Im pretty tired of this "problem". Nobody posts a step-by-step tutorial to reproduce the problem or a demo fiddle, but one time users complain all the time "Oaaah it does not work! Fix it in your free time without any input" and dont read any of the discussion.

Yes. Focus is the problem. It was an initial mistake (when filament group developed this years ago) to use focus but its needed to make pop-up style working. And no, Im not able to reproduce the issue.

**Comment by fnagel** [comment]
Sorry for this but no input for weeks and month is pretty annoying and I wanted to release v1.2. today :-(

**Comment by thg2k** [comment]
I totally understand you. :) give me 10 minutes i'll try to make a fiddle.

**Comment by thg2k** [comment]
Jesus christ! It took me forever but I isolated the problem.. it happens only when the selectmenu is not displayed from the initialization phase, because u call _refreshPosition() inside the timer..so if for example it's inside a dialog (it was my case, that's why you couldn't reproduce it) the first call to refreshposition doesn't do anything and the bug appears.
http://thgnet.it/stuff/data/selectmenu-bug147.html

With this test demo (works both on FF and chrome) I guess the timer call with _refreshposition becomes useless and should be removed, so we can find a more proper solution.
Do you have a test case demonstrating what happens when you simply always call refreshposition before focus?

**Comment by thg2k** [comment]
I solved the problem with this commit on my production fork:
https://github.com/thg2k/jquery-ui/commit/1fdd7f92193aee3928af471b77e572b79eb4e12e

If you want me to prepare a pull request for the current master branch let me know. I don't know if it breaks older browser or the "popup" style, I didn't check because I don't need that.

EDIT: And please reopen this issue because it's not solved yet :)

**Comment by fnagel** [comment]
I did not test it but the way you changed the source Im pretty sure it will break popup style. I tested a very long time with different handling of refreshPosition but without success. My idea is that the problem is related to the position plugin and its handling when an element has no focus. There have been several people trying to provide a fix (see list above) but the popup style is needed.

Would be interesting to see how the new version (#140) works out as its handles the focus completely different. I will give it a try.

Ive tested your demo and, there is a problem. And its related to the focus problem. BUT, its not the same problem which this issue is / was related to. The way Selectmenu is built its not meant to be initiated within hidden elements. Just init it after showing it or try using jQuery Peek (see wiki, I know Im repeating myself :-)
See demo (with error): http://jsfiddle.net/QFwNU/
See demo (without error): http://jsfiddle.net/QFwNU/1/

ps: would you mind using fiddle instead of self hosted files? This makes it much easier to work on a problem together. Heres a working version with latest files: http://jsfiddle.net/fnagel/GXtpC/

**Comment by fnagel** [comment]
I feel cautiously  optimistic Ive found a solution for this whole flickering/jumping/hidden/positioning issue. Finally. Thanks to a demo from @thg2k and some know how Im gained from working on the new Selectmenu widget (see #140).
Demo file (see post above) with latest version: http://jsfiddle.net/QFwNU/2/

**Comment by fnagel** [comment]
Please test the latest commit as Ive changed a few position and focus relevant functions which should make selectmenu work nice again in nearly all circumstance. Ive tested in FF7, IE7 and IE8 (VM), Safari 5.0.5 and latest Chrome.

Fact is we need the focus() as the whole focus/keyboard control persist on setting the focus when opening. But Ive managed to change how the menu is positioned and when the focus is set in a way that does not break the popup style. I could get rid of some flickering in IE, too.

Please test and report!

**Comment by fnagel** [comment]
Another update thanks to feedback at #184 (with positive feedback).

Please test and give feedback!

**Comment by Westbrook** [comment]
I've been seeing what I think is this issue in IE9.  Clicking the selectmenu caused the browser to scroll down passed the element so that in some cases you can even see the list of items to select from.  After a lot of testing I found that if I reordered the .focus() and ._refreshPosition() calls at the end of the "open" function and in my limited use case append no new issues to the page.

Previous code was as follows starting on line 508 of the stable release:

```
this.list.addClass(self.widgetBaseClass + '-open')
    .attr('aria-hidden', false)
    .find('li:not(.' + self.widgetBaseClass + '-group):eq(' + this._selectedIndex() + ') a')[0].focus();
if ( this.options.style == "dropdown" ) {
    this.newelement.removeClass('ui-corner-all').addClass('ui-corner-top');
}
this._refreshPosition();
this._trigger("open", event, this._uiHash());
```

My update is as follows:

```
this.list.addClass(self.widgetBaseClass + '-open')
    .attr('aria-hidden', false);
if ( this.options.style == "dropdown" ) {
    this.newelement.removeClass('ui-corner-all').addClass('ui-corner-top');
}
this._refreshPosition();
this._trigger("open", event, this._uiHash());
this.list.find('li:not(.' + self.widgetBaseClass + '-group):eq(' + this._selectedIndex() + ') a')[0].focus();
```

Not sure if this would cause issues in more complex implementations of the plug in, but it seems to be doing the trick for me.

**Comment by fnagel** [comment]
Did you test with my latest commit? Afaics you did use an older code base. Please check the latest commit which seem to fix any problem.

I assume you code changes would break pop-up style, did you test that?

**Comment by Westbrook** [comment]
You are absolutely right about the pop-up style, it doesn't completely break it, but it does prevent the click away event.  Just tested the newest commit and it prevents the scroll change.  My custom styling isn't forward compatible, but that's my problem to fix.  

Thanks for the great plug-in and all the help with updates!

**Comment by fnagel** [comment]
Nice to hear! 
When you need some custom styling you definitly need to check the format and wrapperElement option, perhaps you could achieve the needed styling without changing the widget itself.

@all
I would need some more feedback to merge the typeAhead patch and finally release v1.2

**Comment by fnagel** [comment]
Please give some feedback!

**Comment by mvanderw** [comment]
Hi there,

I've just started implementing selectmenu today and ran into this issue. I haven't looked into it much yet but just wanted to give you some quick preliminary feedback along with (hopefully) a nice real-world testcase.

The latest commit appears to fix the issue (I've only tested in Chrome) but seems to break the scrollbars inside long menu's. This may be completely unrelated or may be related to my Webkit scrollbar styles- feel free to disregard if so. :-)

I can reproduce the scrolling behaviour reliably at the filter form I'm currently implementing, accessible at http://holocron.so/db. That page uses the production version you're offering; add ?selectmenu=dev to the url to switch to the latest dev version I cloned from here. The two differences I'm seeing are the fixed positioning and the unusable scrollbars.

Sorry I don't have more in-depth feedback right now (late to the party), just wanted to get this comment out quickly today and take the opportunity to thank you for all the work you've clearly put into this plugin. It's much appreciated! =)

**Comment by fnagel** [comment]
Hey @mvanderw,

I've tried your demo in latest Chrome on Win7 and noticed I can not click the scrollbar. The problem does not occur in the demo files so I assume its related to your specific demo. Do you use some color scheme on the scrollbars?
I would agree its not related to this issue, so please open a new one for your scrollbar issue. 

Thanks for the nice demo, would be nice if you could post it here #60 when its ready!

**Comment by fnagel** [comment]
All problems seems to be fixed in v1.2. Please open a new issue if similar problem occur.
